### DISCUSSION TITLE
Custom SQL Queries Feature

### DISCUSSION THREAD

**Discussion by j-ova** [PostgREST/postgrest#2621] [Q&A] [answered]
Hello everyone,

First of all thank you for this amazing project. I only start to grasp what its possibilities are, but the more I find the more I feel confident it will answer most of our needs.

My question relates to custom SQL Queries. My end users know SQL and might want to send custom queries without having to learn how to translate them as HTTP Query Parameters. 
I found no way to send a fully written SQL Query, and the reason seems to be (from the documentation) that "The PostgREST URL grammar limits the kinds of queries clients can perform. It prevents arbitrary, potentially poorly constructed and slow client queries.".

Is this the only reason that feature has never been implemented? Because it seems to me that the raw power and flexibility of such a feature should be enough to implement it and let people decide wether it is safe or not to use on their environment. 

I am not sur yet if this is something I would have time to spend on, but if there are no technical or security limitations to foresee, is it a feature you would accept?

Thank you for your time.
Regards,
J

**Comment by steve-chavez** [comment, marked as answer]
Hello,

> Is this the only reason that feature has never been implemented?

That and it's too easy for malicious users to DoS your db by executing expensive queries.

However, how about solving this with a function that executes dynamic SQL:

```sql
-- note that the function returns bytea, so it can be extended to do arbitrary formats including binary formats
CREATE OR REPLACE FUNCTION test.raw_sql(text) RETURNS SETOF bytea
AS $_$
BEGIN
    RETURN QUERY EXECUTE format($$ WITH pgrst_source AS (%s) SELECT json_agg(pgrst_source)::text::bytea AS body FROM pgrst_source $$, $1);
END; $_$
LANGUAGE plpgsql;
```

```bash
curl 'localhost:3000/rpc/raw_sql' \
  -H "Content-Type: text/plain" -H "Accept: application/octet-stream" \
  -d @- <<SQL

SELECT * FROM test.projects x WHERE id in (1,2)

SQL

[{"id":1,"name":"Windows 7","client_id":1},
 {"id":2,"name":"Windows 10","client_id":1}]
```

(You could hide the required headers behind a proxy for now)

### Caveats:

- No TCL(BEGIN, COMMIT, etc) commands are possible since `EXECUTE` doesn't work with those
- DML(UPDATE, DELETE, INSERT) requires a RETURNING clause at the end due to the use of CTE. Otherwise it would be invalid SQL syntax.
- No DDL(CREATE TABLE, etc) due to the use of CTE

### Advantages

- You can grant access to this function only to trusted users - using a role of your choosing.
- Clear error messages are already available
```bash
curl 'localhost:3000/rpc/raw_sql' \
  -H "Content-Type: text/plain" -H "Accept: application/octet-stream" \
  -d @- <<SQL

INSERT INTO test.projects VALUES (1, 'IOS', null) RETURNING *

SQL

{"code":"23505",
 "details":"Key (id)=(1) already exists.",
 "hint":null,
 "message":"duplicate key value violates unique constraint \"projects_pkey\""}
```

WDYT?

**Reply by wolfgangwalther** [reply]
This will allow the user to call `SET ROLE` to change into any role that has been granted to the `authenticator` role. This will render all of your role management useless.

Don't do it.

**Reply by j-ova** [reply]
Our PostgREST would only be used for read operations for BI purposes and only one role (granted select on all tables and views) is expected to be needed.

So.... I love the suggestion and will try that one as soon as possible. 
Thank you @steve-chavez for the idea, and thank you @wolfgangwalther for the security warning!

**Reply by j-ova** [reply]
Definitely worked as intended. Thank you!

**Reply by steve-chavez** [reply]
> This will allow the user to call SET ROLE to change into any role that has been granted to the authenticator role

@wolfgangwalther Is the above really possible? So first `SET ROLE` won't work inside a CTE, you could do `set_config()` but I don't think that would work in a CTE since the role would have to be set before the query is ran to change the privileges.

Lastly, I'm assuming the role allowed to execute this function is almost superuser-like, so it wouldn't make sense to downgrade to another role.

---

One thing that this does deny is our enforcement of the `search_path` for every request. In this case the user would be able to point to any schema it can access, including `pg_catalog`.
